Hudl.Mjolnir/Command/BreakerInvoker.cs

```diff
@@ -0,0 +1,142 @@
+﻿using Hudl.Mjolnir.Breaker;
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Hudl.Mjolnir.Command
+{
+    internal interface IBreakerInvoker
+    {
+        Task<TResult> ExecuteWithBreakerAsync<TResult>(AsyncCommand<TResult> command, CancellationToken ct);
+        TResult ExecuteWithBreaker<TResult>(SyncCommand<TResult> command, CancellationToken ct);
+    }
+
+    internal class BreakerInvoker : IBreakerInvoker
+    {
+        private readonly ICommandContext _context;
+
+        public BreakerInvoker(ICommandContext context)
+        {
+            _context = context ?? CommandContext.Current;
+        }
+
+        public async Task<TResult> ExecuteWithBreakerAsync<TResult>(AsyncCommand<TResult> command, CancellationToken ct)
+        {
+            var breaker = _context.GetCircuitBreaker(command.BreakerKey);
+
+            if (!breaker.IsAllowing())
+            {
+                _context.MetricEvents.RejectedByBreaker(breaker.Name, command.Name);
+                throw new CircuitBreakerRejectedException();
+            }
+
+            TResult result;
+
+            var success = true;
+            var breakerStopwatch = Stopwatch.StartNew();
+            var executionStopwatch = Stopwatch.StartNew();
+            try
+            {
+                // Await here so we can catch the Exception and track the state.
+                // I suppose we could do this with a continuation, too. Await's easier.
+                result = await command.ExecuteAsync(ct).ConfigureAwait(false);
```

Comment:
This shouldn't affect context around `await`s higher up the stack. As I understand it, `ConfigureAwait(false)` *should* only apply to the `await` it targets, in the frame of the method being `await`ed within (though it's hard to find a clear statement declaring that fact). `await`s higher up the stack will retain context around the call as long as they also don't `ConfigureAwait(false)`.

I think, a while back, I did a test in a simple MVC app and verified that context was retained on higher `await`s when lower `await`s used `ConfigureAwait(false)`, but it's been a while, so I might be mis-remembering what I was testing.